File: TCPIP/src/10/io_uring_tcp_echo_server.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <liburing.h>

#define QUEUE_DEPTH 256
#define BUFFER_SIZE 4096
#define PORT 8080

enum {
    EVENT_ACCEPT,
    EVENT_READ,
    EVENT_WRITE
};

typedef struct {
    int event_type;
    int fd;
    char *buffer;
    size_t len;
} conn_info;

conn_info* create_conn_info(int event_type, int fd) {
    conn_info *info = malloc(sizeof(conn_info));
    info->event_type = event_type;
    info->fd = fd;
    info->buffer = malloc(BUFFER_SIZE);
    info->len = 0;
    return info;
}

void free_conn_info(conn_info *info) {
    if (info->buffer) free(info->buffer);
    free(info);
}

int setup_listening_socket(int port) {
    int sock_fd = socket(AF_INET, SOCK_STREAM, 0);
    if (sock_fd < 0) {
        perror("socket");
        return -1;
    }

    int enable = 1;
    setsockopt(sock_fd, SOL_SOCKET, SO_REUSEADDR, &enable, sizeof(enable));

    struct sockaddr_in srv_addr;
    memset(&srv_addr, 0, sizeof(srv_addr));
    srv_addr.sin_family = AF_INET;
    srv_addr.sin_addr.s_addr = INADDR_ANY;
    srv_addr.sin_port = htons(port);

    if (bind(sock_fd, (struct sockaddr*)&srv_addr, sizeof(srv_addr)) < 0) {
        perror("bind");
        return -1;
    }

    if (listen(sock_fd, 128) < 0) {
        perror("listen");
        return -1;
    }

    return sock_fd;
}

void add_accept_request(struct io_uring *ring, int server_fd, 
                       struct sockaddr_in *client_addr, socklen_t *client_len) {
    struct io_uring_sqe *sqe = io_uring_get_sqe(ring);
    conn_info *info = create_conn_info(EVENT_ACCEPT, server_fd);
    
    io_uring_prep_accept(sqe, server_fd, (struct sockaddr*)client_addr, 
                        client_len, 0);
    io_uring_sqe_set_data(sqe, info);
}

void add_read_request(struct io_uring *ring, int client_fd) {
    struct io_uring_sqe *sqe = io_uring_get_sqe(ring);
    conn_info *info = create_conn_info(EVENT_READ, client_fd);
    
    io_uring_prep_recv(sqe, client_fd, info->buffer, BUFFER_SIZE, 0);
    io_uring_sqe_set_data(sqe, info);
}

void add_write_request(struct io_uring *ring, int client_fd, 
                      char *buffer, size_t len) {
    struct io_uring_sqe *sqe = io_uring_get_sqe(ring);
    conn_info *info = create_conn_info(EVENT_WRITE, client_fd);
    
    memcpy(info->buffer, buffer, len);
    info->len = len;
    
    io_uring_prep_send(sqe, client_fd, info->buffer, len, 0);
    io_uring_sqe_set_data(sqe, info);
}

int main() {
    struct io_uring ring;
    struct io_uring_cqe *cqe;
    struct sockaddr_in client_addr;
    socklen_t client_len = sizeof(client_addr);

    // Initialize io_uring
    if (io_uring_queue_init(QUEUE_DEPTH, &ring, 0) < 0) {
        perror("io_uring_queue_init");
        return 1;
    }

    // Setup listening socket
    int server_fd = setup_listening_socket(PORT);
    if (server_fd < 0) {
        return 1;
    }

    printf("Echo server listening on port %d\n", PORT);

    // Add initial accept request
    add_accept_request(&ring, server_fd, &client_addr, &client_len);
    io_uring_submit(&ring);

    // Event loop
    while (1) {
        int ret = io_uring_wait_cqe(&ring, &cqe);
        if (ret < 0) {
            perror("io_uring_wait_cqe");
            break;
        }

        conn_info *info = (conn_info*)io_uring_cqe_get_data(cqe);
        int result = cqe->res;

        if (result < 0) {
            fprintf(stderr, "Operation failed: %s\n", strerror(-result));
            if (info->event_type != EVENT_ACCEPT) {
                close(info->fd);
            }
            free_conn_info(info);
            io_uring_cqe_seen(&ring, cqe);
            continue;
        }

        switch (info->event_type) {
            case EVENT_ACCEPT: {
                int client_fd = result;
                printf("Accepted new connection: fd=%d\n", client_fd);
                
                // Add read request for new client
                add_read_request(&ring, client_fd);
                
                // Add another accept request
                add_accept_request(&ring, server_fd, &client_addr, &client_len);
                io_uring_submit(&ring);
                break;
            }

            case EVENT_READ: {
                if (result == 0) {
                    // Client closed connection
                    printf("Client closed connection: fd=%d\n", info->fd);
                    close(info->fd);
                } else {
                    printf("Received %d bytes from fd=%d\n", result, info->fd);
                    
                    // Echo back the data
                    add_write_request(&ring, info->fd, info->buffer, result);
                    io_uring_submit(&ring);
                }
                break;
            }

            case EVENT_WRITE: {
                printf("Sent %d bytes to fd=%d\n", result, info->fd);
                
                // Continue reading from this client
                add_read_request(&ring, info->fd);
                io_uring_submit(&ring);
                break;
            }
        }

        free_conn_info(info);
        io_uring_cqe_seen(&ring, cqe);
    }

    close(server_fd);
    io_uring_queue_exit(&ring);
    return 0;
}

// Compile: gcc -o echo_server echo_server.c -luring
// Run: ./echo_server
// Test: telnet localhost 8080

File: TCPIP/src/10/io_uring_tcp_echo_server.rs
// Cargo.toml dependencies:
// [dependencies]
// io-uring = "0.6"
// socket2 = "0.5"

use io_uring::{opcode, types, IoUring};
use socket2::{Domain, Protocol, Socket, Type};
use std::collections::HashMap;
use std::io;
use std::net::SocketAddr;
use std::os::unix::io::{AsRawFd, RawFd};

const QUEUE_DEPTH: u32 = 256;
const BUFFER_SIZE: usize = 4096;
const PORT: u16 = 8080;

#[derive(Debug)]
enum EventType {
    Accept,
    Read,
    Write,
}

struct ConnectionInfo {
    event_type: EventType,
    fd: RawFd,
    buffer: Vec<u8>,
}

impl ConnectionInfo {
    fn new(event_type: EventType, fd: RawFd) -> Self {
        Self {
            event_type,
            fd,
            buffer: vec![0u8; BUFFER_SIZE],
        }
    }
}

fn setup_listening_socket(port: u16) -> io::Result<Socket> {
    let socket = Socket::new(Domain::IPV4, Type::STREAM, Some(Protocol::TCP))?;
    socket.set_reuse_address(true)?;
    
    let addr: SocketAddr = format!("0.0.0.0:{}", port).parse().unwrap();
    socket.bind(&addr.into())?;
    socket.listen(128)?;
    
    println!("Echo server listening on port {}", port);
    Ok(socket)
}

fn main() -> io::Result<()> {
    // Initialize io_uring
    let mut ring = IoUring::new(QUEUE_DEPTH)?;
    let (submitter, mut sq, mut cq) = ring.split();

    // Setup listening socket
    let listener = setup_listening_socket(PORT)?;
    let listener_fd = listener.as_raw_fd();

    // Track active operations
    let mut token_counter: u64 = 0;
    let mut buffers: HashMap<u64, Box<ConnectionInfo>> = HashMap::new();

    // Helper function to get next token
    let mut next_token = || {
        token_counter += 1;
        token_counter
    };

    // Submit initial accept operation
    let token = next_token();
    let mut conn_info = Box::new(ConnectionInfo::new(EventType::Accept, listener_fd));
    let accept_addr = conn_info.buffer.as_mut_ptr() as *mut libc::sockaddr;
    let accept_len = conn_info.buffer.as_mut_ptr().wrapping_add(64) as *mut libc::socklen_t;
    
    unsafe {
        let accept_e = opcode::Accept::new(
            types::Fd(listener_fd),
            accept_addr,
            accept_len,
        )
        .build()
        .user_data(token);
        
        sq.push(&accept_e).expect("Queue is full");
    }
    
    buffers.insert(token, conn_info);
    sq.sync();
    submitter.submit()?;

    // Event loop
    loop {
        // Wait for completion events
        cq.sync();
        
        for cqe in &mut cq {
            let token = cqe.user_data();
            let result = cqe.result();
            
            if let Some(mut info) = buffers.remove(&token) {
                if result < 0 {
                    eprintln!("Operation failed: {}", io::Error::from_raw_os_error(-result));
                    if !matches!(info.event_type, EventType::Accept) {
                        unsafe { libc::close(info.fd) };
                    }
                    continue;
                }

                match info.event_type {
                    EventType::Accept => {
                        let client_fd = result;
                        println!("Accepted new connection: fd={}", client_fd);

                        // Submit read operation for new client
                        let read_token = next_token();
                        let mut read_info = Box::new(ConnectionInfo::new(
                            EventType::Read,
                            client_fd,
                        ));
                        
                        let read_e = opcode::Recv::new(
                            types::Fd(client_fd),
                            read_info.buffer.as_mut_ptr(),
                            BUFFER_SIZE as u32,
                        )
                        .build()
                        .user_data(read_token);
                        
                        unsafe {
                            sq.push(&read_e).expect("Queue is full");
                        }
                        buffers.insert(read_token, read_info);

                        // Submit another accept operation
                        let accept_token = next_token();
                        let mut accept_info = Box::new(ConnectionInfo::new(
                            EventType::Accept,
                            listener_fd,
                        ));
                        
                        let accept_addr = accept_info.buffer.as_mut_ptr() as *mut libc::sockaddr;
                        let accept_len = accept_info.buffer.as_mut_ptr().wrapping_add(64) 
                            as *mut libc::socklen_t;
                        
                        unsafe {
                            let accept_e = opcode::Accept::new(
                                types::Fd(listener_fd),
                                accept_addr,
                                accept_len,
                            )
                            .build()
                            .user_data(accept_token);
                            
                            sq.push(&accept_e).expect("Queue is full");
                        }
                        buffers.insert(accept_token, accept_info);
                    }

                    EventType::Read => {
                        if result == 0 {
                            // Client closed connection
                            println!("Client closed connection: fd={}", info.fd);
                            unsafe { libc::close(info.fd) };
                        } else {
                            let bytes_read = result as usize;
                            println!("Received {} bytes from fd={}", bytes_read, info.fd);

                            // Submit write operation (echo back)
                            let write_token = next_token();
                            let mut write_info = Box::new(ConnectionInfo::new(
                                EventType::Write,
                                info.fd,
                            ));
                            write_info.buffer[..bytes_read]
                                .copy_from_slice(&info.buffer[..bytes_read]);

                            let write_e = opcode::Send::new(
                                types::Fd(info.fd),
                                write_info.buffer.as_ptr(),
                                bytes_read as u32,
                            )
                            .build()
                            .user_data(write_token);

                            unsafe {
                                sq.push(&write_e).expect("Queue is full");
                            }
                            buffers.insert(write_token, write_info);
                        }
                    }

                    EventType::Write => {
                        println!("Sent {} bytes to fd={}", result, info.fd);

                        // Submit another read operation
                        let read_token = next_token();
                        let mut read_info = Box::new(ConnectionInfo::new(
                            EventType::Read,
                            info.fd,
                        ));

                        let read_e = opcode::Recv::new(
                            types::Fd(info.fd),
                            read_info.buffer.as_mut_ptr(),
                            BUFFER_SIZE as u32,
                        )
                        .build()
                        .user_data(read_token);

                        unsafe {
                            sq.push(&read_e).expect("Queue is full");
                        }
                        buffers.insert(read_token, read_info);
                    }
                }
            }
        }

        // Submit all pending operations
        sq.sync();
        submitter.submit_and_wait(1)?;
    }
}

// Build: cargo build --release
// Run: cargo run
// Test: telnet localhost 8080